File: backend/internal/model/feature_flag_audit_log.go
package model

import "time"

// FeatureFlagAuditLog は feature_flag_audit_logs テーブルの写し
// admin/go・admin/railsがフラグを変更するたびに1行追記する「誰が・いつ・何を変更したか」の履歴
// FlagKeyを非正規化して持つのは、対象のFeatureFlag行が削除された後も履歴だけは読めるようにするため
type FeatureFlagAuditLog struct {
	ID                     uint64    `gorm:"column:id;primaryKey"`
	FeatureFlagID          uint64    `gorm:"column:feature_flag_id"`
	FlagKey                string    `gorm:"column:flag_key"`
	BeforeDefaultVariation *string   `gorm:"column:before_default_variation"`
	AfterDefaultVariation  string    `gorm:"column:after_default_variation"`
	BeforeEnabled          *bool     `gorm:"column:before_enabled"`
	AfterEnabled           bool      `gorm:"column:after_enabled"`
	ChangedBy              string    `gorm:"column:changed_by"`
	ChangedAt              time.Time `gorm:"column:changed_at"`
}

func (FeatureFlagAuditLog) TableName() string {
	return "feature_flag_audit_logs"
}

File: admin/rails/app/models/feature_flag_audit_log.rb
# feature_flag_audit_logs も backend が正本として管理するテーブル(FeatureFlag参照)
# flag_keyを非正規化して持っているので、対象のFeatureFlagが後から削除されても
# 履歴だけは読める(このRailsアプリではFeatureFlagの削除機能自体は提供しない想定だが、
# 念のための設計)
class FeatureFlagAuditLog < ApplicationRecord
  self.table_name = "feature_flag_audit_logs"

  belongs_to :feature_flag

  validates :flag_key, presence: true
  validates :after_default_variation, presence: true
  validates :changed_by, presence: true
end
